File: server/internal/usecase/admin/dto/product_catalog.go
package dto

import "time"

type ProductListQuery struct {
	Page    int    `form:"page" validate:"omitempty,min=1"`
	PerPage int    `form:"per_page" validate:"omitempty,min=1,max=100"`
	Keyword string `form:"keyword" validate:"omitempty,max=96"`
	Type    string `form:"type" validate:"omitempty,oneof=server"`
	Status  string `form:"status" validate:"omitempty,oneof=draft active inactive"`
}

type ProductRequest struct {
	ProductNo   string  `json:"product_no" validate:"omitempty,max=64"`
	Type        string  `json:"type" validate:"omitempty,oneof=server"`
	Slug        string  `json:"slug" validate:"required,min=2,max=96"`
	Name        string  `json:"name" validate:"required,min=1,max=128"`
	Summary     *string `json:"summary" validate:"omitempty,max=255"`
	Description *string `json:"description" validate:"omitempty,max=5000"`
	Status      string  `json:"status" validate:"omitempty,oneof=draft active inactive"`
	Visible     bool    `json:"visible"`
	SortOrder   int     `json:"sort_order" validate:"omitempty,min=0,max=100000"`
}

type ProductStatusRequest struct {
	Status string `json:"status" validate:"required,oneof=draft active inactive"`
}

type ProductItem struct {
	ID          uint64    `json:"id"`
	ProductNo   string    `json:"product_no"`
	Type        string    `json:"type"`
	Slug        string    `json:"slug"`
	Name        string    `json:"name"`
	Summary     *string   `json:"summary"`
	Description *string   `json:"description"`
	Status      string    `json:"status"`
	Visible     bool      `json:"visible"`
	SortOrder   int       `json:"sort_order"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

type ProductPlanListQuery struct {
	Page      int    `form:"page" validate:"omitempty,min=1"`
	PerPage   int    `form:"per_page" validate:"omitempty,min=1,max=100"`
	ProductID uint64 `form:"product_id" validate:"omitempty,min=1"`
	Keyword   string `form:"keyword" validate:"omitempty,max=96"`
	Status    string `form:"status" validate:"omitempty,oneof=draft active inactive sold_out"`
}

type ProductPlanRequest struct {
	PlanNo         string  `json:"plan_no" validate:"omitempty,max=64"`
	ProductID      uint64  `json:"product_id" validate:"required,min=1"`
	Code           string  `json:"code" validate:"required,min=2,max=96"`
	Name           string  `json:"name" validate:"required,min=1,max=128"`
	Summary        *string `json:"summary" validate:"omitempty,max=255"`
	CPUCores       int     `json:"cpu_cores" validate:"required,min=1,max=512"`
	MemoryMB       int     `json:"memory_mb" validate:"required,min=128,max=1048576"`
	SystemDiskGB   int     `json:"system_disk_gb" validate:"required,min=1,max=1048576"`
	DataDiskGB     int     `json:"data_disk_gb" validate:"omitempty,min=0,max=1048576"`
	BandwidthMbps  int     `json:"bandwidth_mbps" validate:"required,min=1,max=100000"`
	TrafficGB      *int    `json:"traffic_gb" validate:"omitempty,min=0,max=10000000"`
	PublicIPCount  int     `json:"public_ip_count" validate:"omitempty,min=0,max=1024"`
	Virtualization string  `json:"virtualization" validate:"required,oneof=kvm"`
	Architecture   string  `json:"architecture" validate:"required,oneof=x86_64"`
	IsFeatured     bool    `json:"is_featured"`
	Status         string  `json:"status" validate:"required,oneof=draft active inactive sold_out"`
	Visible        bool    `json:"visible"`
	SortOrder      int     `json:"sort_order" validate:"omitempty,min=0,max=100000"`
}

type ProductPlanStatusRequest struct {
	Status string `json:"status" validate:"required,oneof=draft active inactive sold_out"`
}

type ProductPlanItem struct {
	ID             uint64    `json:"id"`
	PlanNo         string    `json:"plan_no"`
	ProductID      uint64    `json:"product_id"`
	Code           string    `json:"code"`
	Name           string    `json:"name"`
	Summary        *string   `json:"summary"`
	CPUCores       int       `json:"cpu_cores"`
	MemoryMB       int       `json:"memory_mb"`
	SystemDiskGB   int       `json:"system_disk_gb"`
	DataDiskGB     int       `json:"data_disk_gb"`
	BandwidthMbps  int       `json:"bandwidth_mbps"`
	TrafficGB      *int      `json:"traffic_gb"`
	PublicIPCount  int       `json:"public_ip_count"`
	Virtualization string    `json:"virtualization"`
	Architecture   string    `json:"architecture"`
	IsFeatured     bool      `json:"is_featured"`
	Status         string    `json:"status"`
	Visible        bool      `json:"visible"`
	SortOrder      int       `json:"sort_order"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
}

type PlanPriceRequest struct {
	BillingCycle       string  `json:"billing_cycle" validate:"required,oneof=monthly quarterly semi_yearly yearly"`
	PriceCents         uint64  `json:"price_cents" validate:"required,min=1"`
	OriginalPriceCents *uint64 `json:"original_price_cents" validate:"omitempty,min=0"`
	Currency           string  `json:"currency" validate:"required,oneof=CNY"`
	Status             string  `json:"status" validate:"required,oneof=active inactive"`
	SortOrder          int     `json:"sort_order" validate:"omitempty,min=0,max=100000"`
}

type PlanPriceListRequest struct {
	Prices []PlanPriceRequest `json:"prices" validate:"required,dive"`
}

type PlanPriceItem struct {
	ID                 uint64    `json:"id"`
	PlanID             uint64    `json:"plan_id"`
	BillingCycle       string    `json:"billing_cycle"`
	PriceCents         uint64    `json:"price_cents"`
	OriginalPriceCents *uint64   `json:"original_price_cents"`
	Currency           string    `json:"currency"`
	Status             string    `json:"status"`
	SortOrder          int       `json:"sort_order"`
	CreatedAt          time.Time `json:"created_at"`
	UpdatedAt          time.Time `json:"updated_at"`
}

type SalesRegionListQuery struct {
	Keyword string `form:"keyword" validate:"omitempty,max=96"`
	Status  string `form:"status" validate:"omitempty,oneof=active inactive"`
}

type SalesRegionRequest struct {
	RegionNo  string  `json:"region_no" validate:"omitempty,max=64"`
	Code      string  `json:"code" validate:"required,min=2,max=64"`
	Name      string  `json:"name" validate:"required,min=1,max=128"`
	Country   *string `json:"country" validate:"omitempty,max=64"`
	City      *string `json:"city" validate:"omitempty,max=64"`
	Summary   *string `json:"summary" validate:"omitempty,max=255"`
	Status    string  `json:"status" validate:"required,oneof=active inactive"`
	Visible   bool    `json:"visible"`
	SortOrder int     `json:"sort_order" validate:"omitempty,min=0,max=100000"`
}

type SalesRegionItem struct {
	ID        uint64    `json:"id"`
	RegionNo  string    `json:"region_no"`
	Code      string    `json:"code"`
	Name      string    `json:"name"`
	Country   *string   `json:"country"`
	City      *string   `json:"city"`
	Summary   *string   `json:"summary"`
	Status    string    `json:"status"`
	Visible   bool      `json:"visible"`
	SortOrder int       `json:"sort_order"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type ServerOSTemplateListQuery struct {
	Keyword string `form:"keyword" validate:"omitempty,max=96"`
	Status  string `form:"status" validate:"omitempty,oneof=active inactive"`
}

type ServerOSTemplateRequest struct {
	TemplateNo   string  `json:"template_no" validate:"omitempty,max=64"`
	Code         string  `json:"code" validate:"required,min=2,max=96"`
	Name         string  `json:"name" validate:"required,min=1,max=128"`
	OSFamily     string  `json:"os_family" validate:"required,oneof=linux windows bsd"`
	Distribution string  `json:"distribution" validate:"required,min=1,max=64"`
	Version      string  `json:"version" validate:"required,min=1,max=64"`
	Architecture string  `json:"architecture" validate:"required,oneof=x86_64"`
	Summary      *string `json:"summary" validate:"omitempty,max=255"`
	Status       string  `json:"status" validate:"required,oneof=active inactive"`
	Visible      bool    `json:"visible"`
	SortOrder    int     `json:"sort_order" validate:"omitempty,min=0,max=100000"`
}

type ServerOSTemplateItem struct {
	ID           uint64    `json:"id"`
	TemplateNo   string    `json:"template_no"`
	Code         string    `json:"code"`
	Name         string    `json:"name"`
	OSFamily     string    `json:"os_family"`
	Distribution string    `json:"distribution"`
	Version      string    `json:"version"`
	Architecture string    `json:"architecture"`
	Summary      *string   `json:"summary"`
	Status       string    `json:"status"`
	Visible      bool      `json:"visible"`
	SortOrder    int       `json:"sort_order"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
}

type NetworkTypeListQuery struct {
	Keyword string `form:"keyword" validate:"omitempty,max=96"`
	Status  string `form:"status" validate:"omitempty,oneof=active inactive"`
}

type NetworkTypeRequest struct {
	NetworkTypeNo string  `json:"network_type_no" validate:"omitempty,max=64"`
	Code          string  `json:"code" validate:"required,min=2,max=64"`
	Name          string  `json:"name" validate:"required,min=1,max=128"`
	Summary       *string `json:"summary" validate:"omitempty,max=255"`
	Status        string  `json:"status" validate:"required,oneof=active inactive"`
	Visible       bool    `json:"visible"`
	SortOrder     int     `json:"sort_order" validate:"omitempty,min=0,max=100000"`
}

type NetworkTypeItem struct {
	ID            uint64    `json:"id"`
	NetworkTypeNo string    `json:"network_type_no"`
	Code          string    `json:"code"`
	Name          string    `json:"name"`
	Summary       *string   `json:"summary"`
	Status        string    `json:"status"`
	Visible       bool      `json:"visible"`
	SortOrder     int       `json:"sort_order"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`
}

type PlanRelationRequest struct {
	IDs []uint64 `json:"ids" validate:"required,dive,min=1"`
}

type PlanRelationResponse struct {
	PlanID     uint64   `json:"plan_id"`
	RelatedIDs []uint64 `json:"related_ids"`
}

File: admin/src/api/product-catalog.ts
import { http, type ApiEnvelope } from '../utils/request'

import type { PaginatedData } from './admin-user'

export interface ProductItem {
  id: number
  product_no: string
  type: string
  slug: string
  name: string
  summary: string | null
  description: string | null
  status: string
  visible: boolean
  sort_order: number
  created_at: string
  updated_at: string
}

export interface ProductPayload {
  product_no?: string
  type: 'server'
  slug: string
  name: string
  summary?: string | null
  description?: string | null
  status: 'draft' | 'active' | 'inactive'
  visible: boolean
  sort_order: number
}

export interface ProductPlanItem {
  id: number
  plan_no: string
  product_id: number
  code: string
  name: string
  summary: string | null
  cpu_cores: number
  memory_mb: number
  system_disk_gb: number
  data_disk_gb: number
  bandwidth_mbps: number
  traffic_gb: number | null
  public_ip_count: number
  virtualization: string
  architecture: string
  is_featured: boolean
  status: string
  visible: boolean
  sort_order: number
  created_at: string
  updated_at: string
}

export interface ProductPlanPayload {
  plan_no?: string
  product_id: number
  code: string
  name: string
  summary?: string | null
  cpu_cores: number
  memory_mb: number
  system_disk_gb: number
  data_disk_gb: number
  bandwidth_mbps: number
  traffic_gb?: number | null
  public_ip_count: number
  virtualization: 'kvm'
  architecture: 'x86_64'
  is_featured: boolean
  status: 'draft' | 'active' | 'inactive' | 'sold_out'
  visible: boolean
  sort_order: number
}

export interface PlanPriceItem {
  id: number
  plan_id: number
  billing_cycle: string
  price_cents: number
  original_price_cents: number | null
  currency: string
  status: string
  sort_order: number
}

export interface PlanPricePayload {
  billing_cycle: 'monthly' | 'quarterly' | 'semi_yearly' | 'yearly'
  price_cents: number
  original_price_cents?: number | null
  currency: 'CNY'
  status: 'active' | 'inactive'
  sort_order: number
}

export interface SalesRegionItem {
  id: number
  region_no: string
  code: string
  name: string
  country: string | null
  city: string | null
  summary: string | null
  status: string
  visible: boolean
  sort_order: number
}

export interface SalesRegionPayload {
  region_no?: string
  code: string
  name: string
  country?: string | null
  city?: string | null
  summary?: string | null
  status: 'active' | 'inactive'
  visible: boolean
  sort_order: number
}

export interface ServerOsTemplateItem {
  id: number
  template_no: string
  code: string
  name: string
  os_family: string
  distribution: string
  version: string
  architecture: string
  summary: string | null
  status: string
  visible: boolean
  sort_order: number
}

export interface ServerOsTemplatePayload {
  template_no?: string
  code: string
  name: string
  os_family: 'linux' | 'windows' | 'bsd'
  distribution: string
  version: string
  architecture: 'x86_64'
  summary?: string | null
  status: 'active' | 'inactive'
  visible: boolean
  sort_order: number
}

export interface NetworkTypeItem {
  id: number
  network_type_no: string
  code: string
  name: string
  summary: string | null
  status: string
  visible: boolean
  sort_order: number
  created_at: string
  updated_at: string
}

export interface NetworkTypePayload {
  network_type_no?: string
  code: string
  name: string
  summary?: string | null
  status: 'active' | 'inactive'
  visible: boolean
  sort_order: number
}

export async function getProducts(params?: Record<string, unknown>) {
  const response = await http.get<ApiEnvelope<PaginatedData<ProductItem>>>('/products', { params })
  return response.data.data
}

export async function createProduct(payload: ProductPayload) {
  const response = await http.post<ApiEnvelope<ProductItem>>('/products', payload)
  return response.data.data
}

export async function updateProduct(id: number, payload: ProductPayload) {
  const response = await http.put<ApiEnvelope<ProductItem>>(`/products/${id}`, payload)
  return response.data.data
}

export async function updateProductStatus(id: number, status: string) {
  const response = await http.patch<ApiEnvelope<ProductItem>>(`/products/${id}/status`, { status })
  return response.data.data
}

export async function deleteProduct(id: number) {
  await http.delete<ApiEnvelope<null>>(`/products/${id}`)
}

export async function getProductPlans(params?: Record<string, unknown>) {
  const response = await http.get<ApiEnvelope<PaginatedData<ProductPlanItem>>>('/product-plans', { params })
  return response.data.data
}

export async function createProductPlan(payload: ProductPlanPayload) {
  const response = await http.post<ApiEnvelope<ProductPlanItem>>('/product-plans', payload)
  return response.data.data
}

export async function updateProductPlan(id: number, payload: ProductPlanPayload) {
  const response = await http.put<ApiEnvelope<ProductPlanItem>>(`/product-plans/${id}`, payload)
  return response.data.data
}

export async function updateProductPlanStatus(id: number, status: string) {
  const response = await http.patch<ApiEnvelope<ProductPlanItem>>(`/product-plans/${id}/status`, { status })
  return response.data.data
}

export async function deleteProductPlan(id: number) {
  await http.delete<ApiEnvelope<null>>(`/product-plans/${id}`)
}

export async function updatePlanPrices(id: number, prices: PlanPricePayload[]) {
  const response = await http.put<ApiEnvelope<PlanPriceItem[]>>(`/product-plans/${id}/prices`, { prices })
  return response.data.data
}

export async function getPlanPrices(id: number) {
  const response = await http.get<ApiEnvelope<PlanPriceItem[]>>(`/product-plans/${id}/prices`)
  return response.data.data
}

export async function updatePlanRegions(id: number, ids: number[]) {
  const response = await http.put<ApiEnvelope<{ plan_id: number; related_ids: number[] }>>(`/product-plans/${id}/regions`, { ids })
  return response.data.data
}

export async function getPlanRegions(id: number) {
  const response = await http.get<ApiEnvelope<SalesRegionItem[]>>(`/product-plans/${id}/regions`)
  return response.data.data
}

export async function updatePlanOsTemplates(id: number, ids: number[]) {
  const response = await http.put<ApiEnvelope<{ plan_id: number; related_ids: number[] }>>(`/product-plans/${id}/os-templates`, { ids })
  return response.data.data
}

export async function updatePlanNetworkTypes(id: number, ids: number[]) {
  const response = await http.put<ApiEnvelope<{ plan_id: number; related_ids: number[] }>>(`/product-plans/${id}/network-types`, { ids })
  return response.data.data
}

export async function getPlanNetworkTypes(id: number) {
  const response = await http.get<ApiEnvelope<NetworkTypeItem[]>>(`/product-plans/${id}/network-types`)
  return response.data.data
}

export async function getPlanOsTemplates(id: number) {
  const response = await http.get<ApiEnvelope<ServerOsTemplateItem[]>>(`/product-plans/${id}/os-templates`)
  return response.data.data
}

export async function getSalesRegions(params?: Record<string, unknown>) {
  const response = await http.get<ApiEnvelope<SalesRegionItem[]>>('/sales-regions', { params })
  return response.data.data
}

export async function createSalesRegion(payload: SalesRegionPayload) {
  const response = await http.post<ApiEnvelope<SalesRegionItem>>('/sales-regions', payload)
  return response.data.data
}

export async function updateSalesRegion(id: number, payload: SalesRegionPayload) {
  const response = await http.put<ApiEnvelope<SalesRegionItem>>(`/sales-regions/${id}`, payload)
  return response.data.data
}

export async function deleteSalesRegion(id: number) {
  await http.delete<ApiEnvelope<null>>(`/sales-regions/${id}`)
}

export async function getServerOsTemplates(params?: Record<string, unknown>) {
  const response = await http.get<ApiEnvelope<ServerOsTemplateItem[]>>('/server-os-templates', { params })
  return response.data.data
}

export async function createServerOsTemplate(payload: ServerOsTemplatePayload) {
  const response = await http.post<ApiEnvelope<ServerOsTemplateItem>>('/server-os-templates', payload)
  return response.data.data
}

export async function updateServerOsTemplate(id: number, payload: ServerOsTemplatePayload) {
  const response = await http.put<ApiEnvelope<ServerOsTemplateItem>>(`/server-os-templates/${id}`, payload)
  return response.data.data
}

export async function deleteServerOsTemplate(id: number) {
  await http.delete<ApiEnvelope<null>>(`/server-os-templates/${id}`)
}

export async function getNetworkTypes(params?: Record<string, unknown>) {
  const response = await http.get<ApiEnvelope<NetworkTypeItem[]>>('/network-types', { params })
  return response.data.data
}

export async function createNetworkType(payload: NetworkTypePayload) {
  const response = await http.post<ApiEnvelope<NetworkTypeItem>>('/network-types', payload)
  return response.data.data
}

export async function updateNetworkType(id: number, payload: NetworkTypePayload) {
  const response = await http.put<ApiEnvelope<NetworkTypeItem>>(`/network-types/${id}`, payload)
  return response.data.data
}

export async function deleteNetworkType(id: number) {
  await http.delete<ApiEnvelope<null>>(`/network-types/${id}`)
}
